### ISSUE TITLE
ID3v2.4 requires use of synch-safe frameLength values  

### ISSUE THREAD

**Issue by jesseward** [mdlayher/taggolib#3]
Some details at  http://id3.org/id3v2.4.0-structure see section 3.1 and 6.2

Given the following from an APIC header

```
000001a0  64 0d 0a 00 41 50 49 43  00 03 2a 41 00 00 03 69  |d...APIC..*A...i|
```

The frame length is the 4 bytes starting at 0x00 03 2a 41 = int32 207425 = synch safe int32 54593. frameLength is storing 207425, which causes us to Seek much further than intended. 

```
        // If frame is attached picture OR frame is too long for buffer, seek past it
        if bytes.Equal(frameBuf, mp3APICFrame) || frameLength > bufLen {
            // Seek past picture data and continue loop
            if _, err := m.reader.Seek(int64(frameLength), 1); err != nil {
                return err
            }

            continue
        }
```

the parseString function in https://github.com/ascherkus/go-id3/blob/master/src/id3/util.go appears to take this into account

**Comment by jesseward** [comment]
I applied a work-around based on @ascherkus and the go-id3 conversion. Re-ran against a small library with APIC and large (COMM) frames. 100% success rate.

Taggolib details from a single file, where previously failed.

```
jesse@analogbubblebath:~/code/go/src/audiostore$ ./audiostore -s /tmp/scan
DEBUG: frame=TIT2, frameLength = 18
DEBUG : id3v2.4, bytestream = [0 0 0 18] , frameLength (unsynched)=18
============
DEBUG: frame=TPE1, frameLength = 19
DEBUG : id3v2.4, bytestream = [0 0 0 19] , frameLength (unsynched)=19
============
DEBUG: frame=TRCK, frameLength = 6
DEBUG : id3v2.4, bytestream = [0 0 0 6] , frameLength (unsynched)=6
============
DEBUG: frame=TALB, frameLength = 17
DEBUG : id3v2.4, bytestream = [0 0 0 17] , frameLength (unsynched)=17
============
DEBUG: frame=TDRC, frameLength = 6
DEBUG : id3v2.4, bytestream = [0 0 0 6] , frameLength (unsynched)=6
============
DEBUG: frame=TCON, frameLength = 10
DEBUG : id3v2.4, bytestream = [0 0 0 10] , frameLength (unsynched)=10
============
DEBUG: frame=APIC, frameLength = 619784
DEBUG : id3v2.4, bytestream = [0 9 117 8] , frameLength (unsynched)=162440
============
DEBUG: frame=TXXX, frameLength = 22
DEBUG : id3v2.4, bytestream = [0 0 0 22] , frameLength (unsynched)=22
============
DEBUG: frame=TPE2, frameLength = 19
DEBUG : id3v2.4, bytestream = [0 0 0 19] , frameLength (unsynched)=19
============
DEBUG: frame=TXXX, frameLength = 26
DEBUG : id3v2.4, bytestream = [0 0 0 26] , frameLength (unsynched)=26
============
DEBUG: frame=TXXX, frameLength = 35
DEBUG : id3v2.4, bytestream = [0 0 0 35] , frameLength (unsynched)=35
============
DEBUG: frame=TSOP, frameLength = 19
DEBUG : id3v2.4, bytestream = [0 0 0 19] , frameLength (unsynched)=19
============
DEBUG: frame=TIT1, frameLength = 28
DEBUG : id3v2.4, bytestream = [0 0 0 28] , frameLength (unsynched)=28
============
DEBUG: frame=TXXX, frameLength = 17
DEBUG : id3v2.4, bytestream = [0 0 0 17] , frameLength (unsynched)=17
============
DEBUG: frame=TCOM, frameLength = 19
DEBUG : id3v2.4, bytestream = [0 0 0 19] , frameLength (unsynched)=19
============
DEBUG: frame=TXXX, frameLength = 42
DEBUG : id3v2.4, bytestream = [0 0 0 42] , frameLength (unsynched)=42
============
DEBUG: frame=TENC, frameLength = 8
DEBUG : id3v2.4, bytestream = [0 0 0 8] , frameLength (unsynched)=8
============
DEBUG: frame=COMM, frameLength = 2052
DEBUG : id3v2.4, bytestream = [0 0 8 4] , frameLength (unsynched)=1028
============
DEBUG: frame=TPUB, frameLength = 14
DEBUG : id3v2.4, bytestream = [0 0 0 14] , frameLength (unsynched)=14
============
DEBUG: frame=TXXX, frameLength = 38
DEBUG : id3v2.4, bytestream = [0 0 0 38] , frameLength (unsynched)=38
============
DEBUG: {album:Smokers Delight albumArtist:Nightmares On Wax artist:Nightmares On Wax file:/tmp/scan/test.mp3 genre:Trip Hop title:Nights Introlude label:Warp Records year:1995 albumPath:/tmp/scan track:1}
```

git diff of mp3.go. I won't commit this as it isn't pretty.

```
diff --git a/mp3.go b/mp3.go
index 97c2c85..01218db 100644
--- a/mp3.go
+++ b/mp3.go
@@ -41,6 +41,19 @@ type mp3Parser struct {
        xingHeader *mp3XingHeader
 }

+func unSynch(data [4]byte) int32 {
+       size := int32(0)
+       for i, b := range data {
+               if b&0x80 > 0 {
+                       fmt.Println("Size byte had non-zero first bit")
+               }
+
+               shift := uint32(len(data)-i-1) * 7
+               size |= int32(b&0x7f) << shift
+       }
+       return size
+}
+
 // Album returns the Album tag for this stream
 func (m mp3Parser) Album() string {
        return m.tags[tagAlbum]
@@ -315,10 +328,19 @@ func (m *mp3Parser) parseID3v2Frames() error {
                                return err
                        }

+                       fmt.Printf("DEBUG: frame=%s, frameLength = %d\n", frameBuf, frameLength)
+                       if m.id3Header.MajorVersion == 4 && !m.id3Header.Unsynchronization {
+                               b := [4]byte{}
+                               binary.BigEndian.PutUint32(b[:], uint32(frameLength))
+                               frameLength = uint32(unSynch(b))
+                               fmt.Printf("DEBUG : id3v2.4, bytestream = %v , frameLength (unsynched)=%d\n", b, frameLength)
+                       }
+
                        // ID3v2.3+: Skip over frame flags
                        if _, err := m.reader.Seek(2, 1); err != nil {
                                return err
                        }
+                       fmt.Println("============")
                }
```

**Comment by mdlayher** [comment]
Neat!  I didn't do too much digging into the ID3 spec, so I'm glad for the feedback.  My library is pretty much all well-tagged FLAC, and I generate MP3s via `mp3fs`, so I imagine that's why I didn't have any issues.

I'll look into this more in the next couple of days, but I wouldn't mind a PR either!  Thanks again for your time.

**Comment by jesseward** [comment]
The ID3v2 spec appears to be more complex than it really should be. This seems fitting http://blog.codinghorror.com/a-spec-tacular-failure/ . 
Thanks for the note about mp3fs, i hadn't seen this and looks pretty neat.

pull https://github.com/mdlayher/taggolib/pull/4 was submitted

**Comment by mdlayher** [comment]
Great article!  Pretty much sums up my experience mucking through the ID3 documentation to try and figure out what it actually means.

`mp3fs` is pretty great!  I've been using it for about 2 years as my daily driver for MP3 creation.

Thanks again for the PR!

**Comment by mdlayher** [comment]
Closed as of merging #4.
